### ISSUE TITLE
BUG: Brimstone crashes if API requests fail to return

### ISSUE THREAD

**Issue by rootabeta** [rootabeta/Brimstone#4]
On 9 Oct Major, 2023, the NS site experienced intermittent service interruptions. During these interruptions, Brimstone instances crashed with the following error message:

```[RDY] BIRD AFFIRM; PREPARED TO ENGAGE
Exception in thread Thread-3 (track_inbounds):
Traceback (most recent call last):
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\urllib3\connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
                       ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\urllib3\connectionpool.py", line 449, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\urllib3\connectionpool.py", line 444, in _make_request
    httplib_response = conn.getresponse()
                       ^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\http\client.py", line 1374, in getresponse
    response.begin()
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\http\client.py", line 318, in begin
    version, status, reason = self._read_status()
                              ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\http\client.py", line 287, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\requests\adapters.py", line 489, in send
    resp = conn.urlopen(
           ^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\urllib3\connectionpool.py", line 787, in urlopen
    retries = retries.increment(
              ^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\urllib3\util\retry.py", line 550, in increment
    raise six.reraise(type(error), error, _stacktrace)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\urllib3\packages\six.py", line 769, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\urllib3\connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
                       ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\urllib3\connectionpool.py", line 449, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\urllib3\connectionpool.py", line 444, in _make_request
    httplib_response = conn.getresponse()
                       ^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\http\client.py", line 1374, in getresponse
    response.begin()
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\http\client.py", line 318, in begin
    version, status, reason = self._read_status()
                              ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\http\client.py", line 287, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\REDACTED\Documents\Brimstone\Brimstone\brimstone.py", line 21, in track_inbounds
    bogeys = radar.ping()
             ^^^^^^^^^^^^
  File "C:\Users\REDACTED\Documents\Brimstone\Brimstone\NSAPI.py", line 221, in ping
    self.newNations = self.fetch_nations()
                      ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\Documents\Brimstone\Brimstone\NSAPI.py", line 184, in fetch_nations
    nationlist = self.api.getNations(canonicalize(self.region))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\Documents\Brimstone\Brimstone\NSAPI.py", line 94, in getNations
    nations = self.getRegionShard(region, "nations")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\Documents\Brimstone\Brimstone\NSAPI.py", line 82, in getRegionShard
    return self.regionAPI(URL)
           ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\Documents\Brimstone\Brimstone\NSAPI.py", line 22, in regionAPI
    r = requests.get(url, headers={"User-Agent":self.user_agent})
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\requests\api.py", line 73, in get
    return request("get", url, params=params, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\requests\api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\requests\sessions.py", line 587, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\requests\sessions.py", line 701, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\REDACTED\AppData\Local\Programs\Python\Python311\Lib\site-packages\requests\adapters.py", line 547, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
```

Preliminary analysis suggests that Brimstone failed to receive a response back from the server in a timely manner and panicked, causing a fatal crash. This is consistent with site behavior at the time.

Expected behavior: Brimstone logs a warning that data may be stale if an API request takes too long (e.g. 5 seconds) to return, and does not modify the internal list of nations. 
Actual behavior: Brimstone provides the error messages seen above and exits. 
Proposed solution: Look into a timeout parameter for requests and error handling.

**Comment by rootabeta** [comment]
_Should_ be handled in https://github.com/rootabeta/Brimstone/commit/e8b57b6f4af915de45a352aed62a4fcfed5d1e86, awaiting review+merge - Rust handles errors a lot more graciously than Python does. Any error during the HTTP requests to the API should now result in a handled error, rather than causing a fatal one. Will close when merge is completed, unless it resurfaces.

**Comment by rootabeta** [comment]
Addressed in https://github.com/rootabeta/Brimstone/commit/e3bf8d5516894df116e756fb76893fffe8341aad, awaiting review prior to merge

**Comment by rootabeta** [comment]
Closed with official release of Brimstone 3.0
